### PR TITLE
Support decompression of pointers to pointers

### LINKED ISSUE #127: Cannot decompress name which is a pointer to a pointer
At my office people are having [an issue](https://github.com/docker/for-mac/issues/1243) with Docker for Mac 1.13.0. It turns out this is due to it's DNS forwarder returning [incorrect results](https://github.com/djs55/ocaml-dns-forward/issues/56) from it's cache. The error in the results is characterised by a name which should be `foo.com.`, but actually is `.` - the root domain. This only happens when the upstream DNS returns an answer containing a name, which is a pointer to a pointer. Information is being lost when the packet is parsed by `ocaml-dns`.

When `ocaml-dns` parses a name which is a pointer to a pointer, it returns a result of no name.

Let's take a look at one of the packets that causes the problem:

```
00000000: 00000000 00000001 10000001 10000000 00000000 00000001  ......
00000006: 00000000 00000010 00000000 00000000 00000000 00000001  ......
0000000c: 00101100 01100011 01101111 01110111 01100010 01100101  ,cowbe
00000012: 01101100 01101100 00101101 01110000 01110010 01101111  ll-pro
00000018: 01100100 00101101 01100100 01101111 01100011 01101011  d-dock
0000001e: 01100101 01110010 00101101 01110010 01100101 01100111  er-reg
00000024: 01101001 01110011 01110100 01110010 01111001 00101101  istry-
0000002a: 01110110 00110010 00101101 01110000 01110010 01101111  v2-pro
00000030: 01100100 00101101 01110011 01110100 01101111 01110010  d-stor
00000036: 01100001 01100111 01100101 00010001 01110011 00110011  age.s3
0000003c: 00101101 01100001 01110000 00101101 01110011 01101111  -ap-so
00000042: 01110101 01110100 01101000 01100101 01100001 01110011  utheas
00000048: 01110100 00101101 00110010 00001001 01100001 01101101  t-2.am
0000004e: 01100001 01111010 01101111 01101110 01100001 01110111  azonaw
00000054: 01110011 00000011 01100011 01101111 01101101 00000000  s.com.
0000005a: 00000000 00000001 00000000 00000001 11000000 00001100  ......
00000060: 00000000 00000101 00000000 00000001 00000000 00000000  ......
00000066: 00000000 00111011 00000000 00000010 11000000 00111001  .;...9
0000006c: 11000000 01101010 00000000 00000001 00000000 00000001  .j....
00000072: 00000000 00000000 00000000 00000100 00000000 00000100  ......
00000078: 00110110 11100111 11111011 00001001 00000000 00000000  6.....
0000007e: 00101001 00001111 10100000 00000000 00000000 00000000  ).....
00000084: 00000000 00000000 00000000                             ...
```

At position 12, there is a sequence of labels:

12: `cowbell-prod-docker-registry-v2-prod-storage`
57: `s3-ap-southeast-2`
75: `amazonaws`
85: `com`

At position 106 there is a pointer to 57:

```
11000000 00111001  .9
```

And at position 108 there is a pointer to 106:

```
11000000 01101010 .j
```

I was hoping to find that section [4.1.4](https://tools.ietf.org/html/rfc1035#section-4.1.4) (message compression) of RFC 1035 would provide some justification for implementing pointers to pointers. Sadly it makes no mention of them.

So I went looking for some DNS test suites and found [this](https://lists.isc.org/pipermail/bind-users/2006-June/062945.html):

> In practice, what BIND does is 'correct'. Or inversely, if you do anything
other than BIND, you are asking for trouble. The RFCs are of secondary
importance, especially where they grant latitude.

and this:

> We mostly validate against BIND and DJBDNS, those being the 'industry standard'.

I can confirm that `dig` and `dnsqr` are able to successfully interpret the pointer to a pointer.  Here's how DJBDNS [does it](https://github.com/abh/djbdns/blob/df36440a03d3aa13996bfc5a0ac87b3a6c0c28d3/dns_packet.c#L35-L77).

@hannesm [notes](https://github.com/djs55/ocaml-dns-forward/issues/56#issuecomment-277989757) that there is a risk if this is implemented without consideration of malicious packets. An acceptable solution should not loop infinitely when encountering a self referential pointer. CERT's [VU#23495](https://www.kb.cert.org/vuls/id/23495) confirms this:

> Numerous DNS implementations do not perform sufficient input checking on compressed labels, allowing an attacker to craft a DNS query containing malformed compressed labels. For example, a packet can be constructed that has a pointer that points back to itself, or that contains a pointer that points to another pointer which points back to the first pointer. When vulnerable programs attempt to decode these malformed labels, various undesirable effects can be induced in the decoding process including allocating all of the process's available memory or causing the process to enter an infinite loop.

Would you be willing to accept a pull request that allows `ocaml-dns` to decompress pointers to pointers? I'm keen to help get this fixed and would be happy to work with you to do so.

### PR DESCRIPTION
**Pull request by yeungda-rea**
The caller of `Name.parse` is required to keep a hash table of memoized names at their offsets whilst parsing a packet. In addition to caching names at label offsets, we now cache names at pointer offsets. This fixes #127.

This change includes three commits. The first addresses the issue. The second attempts to make similar failures explicit. The third adds unit test coverage for the name parsing.

If there is anything further I can do to help resolve this issue, please let me know.

Thank you for taking the time to look over this PR. Here is a kitten in a sink.

thirsty-kitten

### CODE REVIEW

File: lib/name.ml [hunk 1]
```diff
@@ -93,15 +93,16 @@ let parse names base buf = (* what. a. mess. *)
           aux (o :: offsets) (n :: name) (base+offset) (Cstruct.shift buf offset)
 
       | (P (p, _), offset) ->
-          let ns = (Hashtbl.find_all names p
-                       |> List.filter (function L _ -> true | _ -> false))
-          in
+          let all = Hashtbl.find_all names p in
+          if ((List.length all) == 0) then
```
Comment (hannesm) [inline lib/name.ml]: you don't need the parenthesis around the condition, and `List.length` is an expensive operation.

What about:
```OCaml
(match all with
 | [] -> failwith (sprintf "Name.parse_pointer...")
 | _ -> ...rest of body...)
```

**Comment by hannesm** [comment]
looks good to me otherwise.

File: lib/name.ml (hunk 1 above)
Comment (talex5) [inline lib/name.ml]: And then you don't need the parentheses around the `match` either...

Comment (hannesm) [inline lib/name.ml]: I suspect you do since this is in a `match` (and there are more cases of the outer match following)

**Comment by djs55** [comment]
I made a small patch following @hannesm's suggestion and have taken the liberty of pushing it to your branch, @yeungda-rea, hope this is ok.

Thanks again for the fix and the tests!

**Comment by yeungda-rea** [comment]
Thanks for your help getting this change up to scratch @hannesm. @djs55, it's appreciated. Your patch saved us a day of waiting.

Anything else I can do to help?

**Comment by hannesm** [comment]
LGTM

**Comment by djs55** [comment]
LGTM also!

**Comment by djs55** [comment]
I've made a PR to update the changelog and release a new version for Mirage 3: https://github.com/mirage/ocaml-dns/pull/130

@yeungda-rea Separately I've backported this to the branch used by vpnkit and have updated vpnkit in the Docker for Mac/Windows master branch. Tests are currently running and hopefully when they pass the build will be available for download and test -- I'll keep you informed.

**Comment by yeungda-rea** [comment]
Fantastic news! I'm looking forward to testing it.

**Comment by djs55** [comment]
@yeungda-rea a nightly channel build is available from https://download.docker.com/mac/nightly/Docker.dmg -- let me know what happens. For reference my "About Docker" shows
```
Version 1.13.1-beta42 (15371)
Channel: nightly
37fe8c8a4c
```
The process for publishing these is a bit new so it's possible I might have messed up and omitted a patch by accident.

**Comment by yeungda-rea** [comment]
I confirm that the pointers to pointers are being resolved in the nightly build!

With this docker for mac version:

```
Version 1.13.1-beta42 (15371)
Channel: nightly
37fe8c8a4c
```

And this DNS configuration:

```
$ cat ~/Library/Containers/com.docker.docker/Data/database/com.docker.driver.amd64-linux/slirp/dns
# { Addresses: 10.129.148.30, Order: 200000; Zones:  }
nameserver 10.129.148.30
order 200000
timeout 2000
```

I tested two queries for the troublesome A record:

```
$ screen ~/Library/Containers/com.docker.docker/Data/com.docker.driver.amd64-linux/tty
/ # dig cowbell-prod-docker-registry-v2-prod-storage.s3-ap-southeast-2.amazonaws
.com

; <<>> DiG 9.10.4-P6 <<>> cowbell-prod-docker-registry-v2-prod-storage.s3-ap-southeast-2.amazonaws.com
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 8372
;; flags: qr rd ra; QUERY: 1, ANSWER: 2, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4000
;; QUESTION SECTION:
;cowbell-prod-docker-registry-v2-prod-storage.s3-ap-southeast-2.amazonaws.com. IN A

;; ANSWER SECTION:
cowbell-prod-docker-registry-v2-prod-storage.s3-ap-southeast-2.amazonaws.com. 50 IN CNAME s3-ap-southeast-2.amazonaws.com.
s3-ap-southeast-2.amazonaws.com. 2 IN   A       52.95.131.13

;; Query time: 41 msec
;; SERVER: 192.168.65.1#53(192.168.65.1)
;; WHEN: Thu Feb 16 00:05:26 UTC 2017
;; MSG SIZE  rcvd: 135

/ # dig cowbell-prod-docker-registry-v2-prod-storage.s3-ap-southeast-2.amazonaws
.com

; <<>> DiG 9.10.4-P6 <<>> cowbell-prod-docker-registry-v2-prod-storage.s3-ap-southeast-2.amazonaws.com
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 1731
;; flags: qr rd ra; QUERY: 1, ANSWER: 2, AUTHORITY: 0, ADDITIONAL: 0

;; QUESTION SECTION:
;cowbell-prod-docker-registry-v2-prod-storage.s3-ap-southeast-2.amazonaws.com. IN A

;; ANSWER SECTION:
cowbell-prod-docker-registry-v2-prod-storage.s3-ap-southeast-2.amazonaws.com. 50 IN CNAME s3-ap-southeast-2.amazonaws.com.
s3-ap-southeast-2.amazonaws.com. 2 IN   A       52.95.131.13

;; Query time: 0 msec
;; SERVER: 192.168.65.1#53(192.168.65.1)
;; WHEN: Thu Feb 16 00:05:27 UTC 2017
;; MSG SIZE  rcvd: 124

```

We can tell by the second query's time of 0msec that the answer came from the cache. We can see that the ANSWER SECTION of both queries is printed the same.

I inspected the packets and the first answer did indeed have the pointer to the pointer.

**Pull request merged**